File: web/pages/site.master.cs
//Copyright 2011 Cloud Sidekick
// 
//Licensed under the Apache License, Version 2.0 (the "License");
//you may not use this file except in compliance with the License.
//You may obtain a copy of the License at
// 
//    http://www.apache.org/licenses/LICENSE-2.0
// 
//Unless required by applicable law or agreed to in writing, software
//distributed under the License is distributed on an "AS IS" BASIS,
//WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//See the License for the specific language governing permissions and
//limitations under the License.
//
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using System.Web.UI;
using System.Web.UI.WebControls;
using System.Web.Security;
using System.Data;
using System.Xml.Linq;
using System.Xml.XPath;
using Globals;

namespace Web
{
    public partial class sitemaster : System.Web.UI.MasterPage
    {
        acUI.acUI ui = new acUI.acUI();
        public acUI.AppGlobals ag = new acUI.AppGlobals();
		
        public string sUserID = "";
        protected void Page_Init(object sender, System.EventArgs e)
        {
            //check if the user is logged in.
            sUserID = ui.GetSessionUserID();
            if (string.IsNullOrEmpty(sUserID)) ui.Logout("");
        }

        protected void Page_Load(object sender, EventArgs e)
        {
			if (!Page.IsPostBack)
            {
				/*
                 * NOW! this app has a lot of pages, and each one needs to check whether or not the current user
                 * has the 'privilege' of viewing this page.
                 * 
                 * Rather than put code on each page to check the role, we'll just put it here and look 
                 * in a master list.
                 * */
				
				//this does a response.redirect.  Wont' work from ajax, we'll have to do something different.
                ui.IsPageAllowed("");
				
				//log page views if logging is enabled
				//this won't work from a web method either, it needs the Page object.  Figure it out.
				ui.addPageViewLog();


			}
        }
    }
}

File: web/static/script/site.master.js
//Copyright 2012 Cloud Sidekick
// 
//Licensed under the Apache License, Version 2.0 (the "License");
//you may not use this file except in compliance with the License.
//You may obtain a copy of the License at
// 
//    http://www.apache.org/licenses/LICENSE-2.0
// 
//Unless required by applicable law or agreed to in writing, software
//distributed under the License is distributed on an "AS IS" BASIS,
//WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//See the License for the specific language governing permissions and
//limitations under the License.
//

$(document).ready(function () {
    //NOTE: this is the main jQuery function that will execute
    //when the DOM is ready.  Things you want defined on 'page load' should 
    //go in here.

    //use this to define constants, set up jQuery objects, etc.

    $("#main-menu").load("uiMethods/wmGetMenu", function() {
	    $("ul.sf-menu").supersubs({
	        minWidth: 18,   // minimum width of sub-menus in em units 
	        maxWidth: 27,   // maximum width of sub-menus in em units 
	        extraWidth: .5     // extra width can ensure lines don't sometimes turn over 
	        // due to slight rounding differences and font-family 
	    }).superfish();  // call supersubs first, then superfish, so that subs are 
	    // not display:none when measuring. Call before initialising 
	    // containing tabs for same reason.
	});

	getCloudAccounts();
	
    //note the selected one for other functions
    $("#header_cloud_accounts").change(function () {
    	$.cookie("selected_cloud_account", $(this).val());
    	$.cookie("selected_cloud_provider", $("#header_cloud_accounts option:selected").attr("provider"));
    	
        if (typeof CloudAccountWasChanged == 'function') {
	    	CloudAccountWasChanged();
        }
    });
});


// Menu
$(function () {
    var tabContainers = $('div.tabs > div');
    tabContainers.hide().filter(':home').hide();
    $('div.tabs ul.tabNavigation a').click(function () {
        tabContainers.slideUp();
        tabContainers.filter(this.hash).slideDown();
        //$('div.tabs ul.tabNavigation a').removeClass('selected');
        //$(this).addClass('selected');
    }).filter(':first').click();
});
// End Menu

//the timer that keeps the heartbeat updated
window.setInterval(updateHeartbeat, 120000);

//and the jquery ajax it performs
function updateHeartbeat() {
    $.ajax({
        type: "GET",
        url: "uiMethods/wmUpdateHeartbeat",
        contentType: "application/json; charset=utf-8",
        dataType: "json",
        success: function (msg) {
            if (msg) {
                showAlert(msg);
            }
        },
        error: function (response) {
        	// if the heartbeat failed, it's likely the server is just gone.
        	// so, close down the gui and show a pretty message, with a link to the login page.

        	// this is a wicked trick for erasing thie browser history to prevent the user
        	// from going back and seeing broken pages.
        	var backlen=history.length;
        	history.go(-backlen);
        	
			msg = '<div class="ui-widget-content ui-corner-all" style="margin-left: auto; margin-right: auto; padding: 20px; width: 500px;">' + 
				'<div class="ui-state-highlight ui-corner-all" style="padding: 10px;">' + 
				'<span class="ui-icon ui-icon-info" style="float: left; margin-right: .3em;"></span><strong>Uh oh...</strong>' + 
				'<br><br>' +
				'<p style="margin-left: 10px;">It appears the server is no longer available.</p><br>' + 
				'<p style="margin-left: 10px;">Please contact an Administrator.</p><br>' + 
				'<p style="margin-left: 10px;"><a href="/static/login.html">Click here</a> to return to the login page.</p>' + 
				'</div>' + 
				'</div>'
        	$("body").html(msg);
        }
    });
}

function getCloudAccounts() {
	// NOTE: This is not async for a reason - other 'page load' ajax calls depend on it.
    $.ajax({
        type: "POST",
        async: false,
        url: "uiMethods/wmGetCloudAccountsForHeader",
        data: '{"sSelected":"' + $("#selected_cloud_account").val() + '"}',
        contentType: "application/json; charset=utf-8",
        dataType: "html",
        success: function (response) {
            $("#header_cloud_accounts").empty().append(response)
        	$.cookie("selected_cloud_account", $("#header_cloud_accounts option:selected").val());
	    	$.cookie("selected_cloud_provider", $("#header_cloud_accounts option:selected").attr("provider"));
        },
        error: function (response) {
            showAlert(response.responseText);
        }
    });
}

function registerCato()
{
	//update the setting
	updateSetting("general","register_cato","registered");
	
	//open the form
	openWindow('http://community.cloudsidekick.com/register-cato?utm_source=cato_app&amp;utm_medium=menu&amp;utm_campaign=app', 'ask', 'location=no,status=no,scrollbars=yes,resizable=yes,width=800,height=700');
	
	//this might not be visible, but try to remove it anyway.
	$("#registercato").remove();
}